### PR TITLE
Refactor: serde

### PR DESCRIPTION
**Pull request by calebmer**

### CODE REVIEW

**Comment by calebmer** [comment]
I'm not actually ready to merge this yet.

**Comment by calebmer** [comment]
Ok, review this now. There are some rough edges that I want to get your feedback on _cough_ `de.rs` _cough_. Good news though, no more build script!

**Comment by svmnotn** [comment]
couch ?

**Comment by calebmer** [comment]
Fixed

File: src/schema/de.rs [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+        }
+      }
+    }
+
+    deserializer.deserialize_map(SchemaVisitor)
+  }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (svmnotn) [inline src/schema/de.rs:240]: More tests.

File: src/schema/definition.rs [hunk 2]
```diff
@@ -10,25 +18,45 @@ use value::Key;
...
-  pub fn add_type(&mut self, type_: Type) {
-    self.types.push(type_);
+  pub fn add_type<K>(&mut self, name: K, type_: Type) where K: Into<Key> {
+    self.types.insert(name.into(), type_);
   }
 
   /// Gets type of a certain name.
-  pub fn find_type<N>(&self, name: &N) -> Option<&Type> where N: PartialEq<Key> {
-    self.types.iter().find(|type_| name.eq(type_.name()))
+  pub fn get_type(&self, name: &Key) -> Option<&Type> {
```
Comment (svmnotn) [inline src/schema/definition.rs]: why not `where K: Into<Key>`?

File: src/schema/definition.rs [hunk 3]
```diff
@@ -42,17 +70,14 @@ impl PartialEq<Definition> for Definition {
 /// Represents a high-level database type.
 #[derive(Debug)]
 pub struct Type {
-  /// The name of the custom type.
-  name: Key,
   /// The schema used to validate data which claims to be of this type.
   schema: Option<Box<Schema + 'static>>
```
Comment (svmnotn) [inline src/schema/definition.rs]: you keep adding `'static` and I'm not entirely sure this will end well.

File: src/value.rs [hunk 4]
```diff
@@ -77,40 +84,59 @@ impl Value {
...
+      Value::I64(value) => serializer.serialize_i64(value),
+      Value::F64(value) => serializer.serialize_f64(value),
+      Value::String(ref value) => serializer.serialize_str(&value),
+      Value::Array(ref value) => value.serialize(serializer),
+      Value::Object(ref value) => value.serialize(serializer)
+    }
+  }
+}
+
+impl Deserialize for Value {
```
Comment (svmnotn) [inline src/value.rs:117]: Isn't this also `impl` in `de`?

File: src/schema/de.rs (hunk 1 above)
Comment (calebmer) [inline src/schema/de.rs:240]: 😖

File: src/schema/definition.rs (hunk 2 above)
Comment (calebmer) [inline src/schema/definition.rs]: I can try, but since we only get a reference it may not work.

File: src/schema/definition.rs (hunk 3 above)
Comment (calebmer) [inline src/schema/definition.rs]: `'static` is required for the `Schema` trait object.

Comment (calebmer) [inline src/schema/definition.rs]: I should probably make a type alias…

File: src/value.rs (hunk 4 above)
Comment (calebmer) [inline src/value.rs:117]: No, not `impl for Value`. `de` `impl`s `Deserialize` for the structs and stuff defined in the `schema` module.

File: src/schema/definition.rs (hunk 2 above)
Comment (calebmer) [inline src/schema/definition.rs]: A variation of that worked. Don't know if this builds with the parameters we'd expect. We'll see when we get there…

**Comment by svmnotn** [comment]
`de` needs more tests, and I'm worried about the `'static` bound on the `Schema` objects.

**Comment by calebmer** [comment]
Did you read the stackoverflow [question](http://stackoverflow.com/a/26213294/1568890) I sent you a while back?

**Comment by calebmer** [comment]
And a big part of `de`s tests is in `definition.rs`. I guess it makes sense to move it back to `de.rs`…

**Comment by calebmer** [comment]
Added more tests, and after reading the SO answer, are you still worried about `'static`? If you need more assurance, [valico](https://github.com/rustless/valico) is doing a very similar thing to our `Schema` type with their `Validator` type. And look [here](https://github.com/rustless/valico/blob/master/src/json_schema/validators/mod.rs#L131) + [here](https://github.com/rustless/valico/search?utf8=%E2%9C%93&q=BoxedValidator).

**Comment by svmnotn** [comment]
Could use some more tests but it's fine for now.

**Comment by calebmer** [comment]
So merge?

**Pull request merged**